# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmTypes.cpp
# Commit: 1e72f84b4206
# Full Hash: 1e72f84b420612e65ef17b820faeb1b6ad41e4b0
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-10-30 03:48:30
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 5: Remove remaining uses of Frame::tls. r=lth
#   
#   Here we remove remaining uses of Frame::tls. There are many places where
#   we use it: in debug frames, in profiling frames, in jit activation, etc.
#   All these places require short fixes to use our new scheme for getting
# ==============================================================================

diff -r 25281e1775c9 -r 1e72f84b4206 js/src/wasm/WasmTypes.cpp
--- a/js/src/wasm/WasmTypes.cpp	Fri Oct 23 18:17:37 2020 +0000
+++ b/js/src/wasm/WasmTypes.cpp	Fri Oct 23 18:17:38 2020 +0000
@@ -601,10 +601,11 @@
 
 /* static */
 DebugFrame* DebugFrame::from(Frame* fp) {
-  MOZ_ASSERT(fp->instance()->code().metadata().debugEnabled);
+  MOZ_ASSERT(
+      GetNearestEffectiveTls(fp)->instance->code().metadata().debugEnabled);
   auto* df =
       reinterpret_cast<DebugFrame*>((uint8_t*)fp - DebugFrame::offsetOfFrame());
-  MOZ_ASSERT(fp->instance() == df->instance());
+  MOZ_ASSERT(GetNearestEffectiveTls(fp)->instance == df->instance());
   return df;
 }
 
@@ -625,6 +626,10 @@
 #endif
 }
 
+Instance* DebugFrame::instance() const {
+  return GetNearestEffectiveTls(&frame_)->instance;
+}
+
 GlobalObject* DebugFrame::global() const {
   return &instance()->object()->global();
 }